### DISCUSSION TITLE
Path parameters support in swagger

### DISCUSSION THREAD

**Discussion by saurabhgharatpara** [dymmond/esmerald#369] [Q&A] [answered]
Is there a way to add path parameters in Gateway route or request method for Swagger. Right now if you create a manual json or yaml file for swagger, It shows up but it does not detect path parameters if it is defined in path. 
For e.g if path is  "/pet/{petId}" . petId cannot be entered

**Comment by tarsil** [comment]
Hi, I'm not sure of I understood you question properly but basically if I did, then what you need to do is simply, for example:

@get("/pet/{petid}")
async def get_pet(petid: str) -> JSONResponse:
    # add your logic here to return

And then in the Gateway, you simply do:

Gateway(handler=get_pet)

All of this is in the docs thoroughly explained  as well 🙂. 

Esmerald automatically detects the path params. When you create a path param in the URL, you must pass it in the handler with the corresponding type.

**Reply by saurabhgharatpara** [reply]
I've attached screenshot of swagger UI demo
They allow for path parameters to be entered . in this case petId
But when i implement this in esmerald Api as given above , it is not allowing to enter petId. basically the parameter field is shown empty
![swagger-screenshot](https://github.com/user-attachments/assets/2b1edd78-8405-4145-ac2f-d3f8b3c11ef8)

**Reply by tarsil** [reply]
I understand so you did implement like I said, right? Ok, did you press that button there "Try it out"? That is from swagger and you need to press it first

**Reply by saurabhgharatpara** [reply]
the screenshot that ive attached is from swagger demo 
https://petstore.swagger.io/#/pet/getPetById


When i implement in esmerald. It is showing as No Parameters
![swagger-screenshot-2](https://github.com/user-attachments/assets/c10a0daf-1217-41b0-93a1-9fa3f60a1fad)

**Reply by tarsil** [reply]
Can you share your code where you define the Gateway and the handler, please?

**Reply by saurabhgharatpara** [reply]
Ok I've found the solution. So basically i needed to include the path parameter as an argument to function. It's working now. Thanks

**Reply by tarsil** [reply, marked as answer]
@saurabhgharatpara Yes, I mentioned that before

```python
from esmerald import Esmerald, get, JSONResponse

# For handler

@get("/pet/{petid}")
async def get_pet(petid: str) -> JSONResponse:
    # add your logic here to return

# And then in the Gateway, you simply do:
Gateway(handler=get_pet)
```

Every path parameter and the type must be declared in the function handler in order to make sure it will be recognised.

**Reply by saurabhgharatpara** [reply]
I also want to ask about adding query parameters in swagger ?  We can access query params by using 
pet_name= request.query_params.get('pet_name')
But it is not visible in swagger UI
It is visible only when i add  "pet_name" as an argument to function and to path 

@get(summary="Retrieve the list of dog",
status_code=status.HTTP_200_OK)
async def get_pets(request: Request,
                                 headers: dict, logger: Any,
                                 pet_name:str="",
                                ) -> Response:

also 
Gateway(handler=get_pets, path="/pet?pet_name={pet_name}"),

Is there any other way or this is correct ?

**Reply by tarsil** [reply]
@saurabhgharatpara so, you are correct, for the query_params to be visible in the swagger UI you will need to declare those in the function as parameters as well. You don't need to declare that URL in the Gateway() as you did, when you put them in the the function as parameters it should be enough.


One thing to consider is the fact that if the query paramarers are not mandatory, then you **must** specificy the parameters as optional using the typing.Optional or the typing.Union[the type, None] or pass a default 

Does this make sense?

**Reply by saurabhgharatpara** [reply]
if i change the gateway as 
Gateway(handler=get_pets, path="/pet")

and function as 
@get(summary="Retrieve the list of dog",
status_code=status.HTTP_200_OK)
async def get_pets(request: Request,
headers: dict, logger: Any,
pet_name:str="",
) -> Response:

Its not visible in the swagger UI

**Reply by tarsil** [reply]
> if i change the gateway as Gateway(handler=get_pets, path="/pet")
> 
> and function as @get(summary="Retrieve the list of dog", status_code=status.HTTP_200_OK) async def get_pets(request: Request, headers: dict, logger: Any, pet_name:str="", ) -> Response:
> 
> Its not visible in the swagger UI

Hmm maybe a regression happened before. I will have a look and get back to you. In the meantime use the way you mentioned before. Thank you for spotting this

**Comment by tarsil** [comment]
@saurabhgharatpara this was a small regression introduced by a big feature 3 versions ago but its being addressed here https://github.com/dymmond/esmerald/pull/371

Soon (today or tomorrow) this should be released and no issues like this should happen again 👍🏼

**Comment by tarsil** [comment]
@saurabhgharatpara version 3.3.4 of Esmerald was released and this query parameters should not be an issue anymore.

There is also the documentation for it if you want https://www.esmerald.dev/extras/query-params/

**Reply by saurabhgharatpara** [reply]
query parameters are visible but there is another issue with path parameters in Swagger

For create function my parameters are 

async def add_pet(pet_db:str,
request: Request,
headers: dict, logger: Any,
) -> Response:

it is taking 'pet_db' as a path parameter but ive not added "pet_db" as a path parameter in path
Gateway(handler=add_pet, path="/add-pet")

**Reply by tarsil** [reply]
When I run your example, I have this:

![image](https://github.com/user-attachments/assets/79f41e02-0665-4e9d-b093-b5780c317504)

This is correct since `pet_db` is a query_param (not declared in the path_param, see in the swagger saying "query"). If you want as a path param, you can do one of two ways.

1. In the handler

![image](https://github.com/user-attachments/assets/a3846337-ef15-4724-97ac-4aa0d1553b6b)

And it becomes like this:

![image](https://github.com/user-attachments/assets/5214de26-276e-48c6-a459-1fd091df5790)

See that now says "path"?

2. The other way is **not in the handler** but in the Gateway.

![image](https://github.com/user-attachments/assets/5a2c973f-c11b-4add-8ce7-8521149415db)

That will have the same result:

![image](https://github.com/user-attachments/assets/d8b1e09a-9ea6-4c19-858d-143fcf5789ae)


So all working as it was supposed to

**Reply by tarsil** [reply]
The swagger differentiates what is a path param and what is a query param under the param name and says "query" or "path"

**Reply by tarsil** [reply]
make sure you have always the latest Esmerald btw, just in case

**Reply by tarsil** [reply]
Btw my urls have "/accounts/..." because of some of my internals but its the "/pets" thing we are talking and that is what your example reflects. I hope it was clear.

Swagger sometimes can be confusing

**Reply by tarsil** [reply]
@saurabhgharatpara does this make sense?

**Reply by saurabhgharatpara** [reply]
yes it makes sense, also i have one more thing to ask we have additional operators like "__eq, __neq, __ge,  __gte" added to query params how can we implement that ?
for e.g. name__eq, name__neq where name is the query param.

**Reply by tarsil** [reply]
Ok, so these are names for operators, basically but it's not a standard for OpenAPI. What I would do? I would use a dict for this and then pass if there. Something like this, for example:

@get(...)
async def example(q: typing.Dict[str, str]): ...

This will enable the OpenAPI for this and where each key of the dict would be the operators you want. Esmerald recognises the dict (object in swagger) and parses for you and you then treat the "q" or whatever you call as a normal python dict. 

Again, these operators are not swagger standard or any standard for query parameters but you can have your own implementation of you need to

**Reply by saurabhgharatpara** [reply]
ok thanks. one issue that im facing is that ive added dependency to function as an argument but it is treated as a query param for swagger
async def example(example_db:str):

**Reply by tarsil** [reply]
Dependencies in Esmerald are clear 🙂. You need the Inject() object and dependencies declared. Why like this? Makes the code clear.

Have a look at the docs: https://esmerald.dymmond.com/dependencies/

You can inject the function as you want by using the dependencies with Inject/Injects

**Reply by tarsil** [reply]
All good @saurabhgharatpara ?

**Reply by saurabhgharatpara** [reply]
There is one issue with swagger, I have one handler function get_badges for fetching list or details based on the endpoint . There are two endpoints 
    Gateway(handler=get_all_badges, path="/badges"),
    Gateway(handler=get_all_badges, path="/badges/{badge_id}"),
 and this is the function 
 async def get_all_badges(request: Request, badge_id: Optional[str]) -> Response:
 
 The issue is that, im getting the same result for both API's , it works in POSTMAN but not in swagger . i think its not considering badge_id

**Reply by tarsil** [reply]
The reason for that @saurabhgharatpara its because of the operation id, they are the same because you have the same function called `get_all_badges` which I understand it and its ok and it works. This might be a swagger level of bug that I can look at it soon but try to add a different operation_id and let me know.

If it works in postman means Esmerald is ok, time to look at swagger.

Btw, I assume the dependency injection was addressed and understood.

EDIT: It was a swagger bug in the operation ids, not your code. You are doing well 👍🏼

**Reply by tarsil** [reply]
@saurabhgharatpara We will be releasing soon something that will also address this on a Gateway level and fixes this OpenAPI operation id situation.

**Reply by tarsil** [reply]
@saurabhgharatpara This should be addressed now in the version 3.3.6 of Esmerald and now the OpenAPI docs should be able to automatically parse to the right operation_id.

**Reply by saurabhgharatpara** [reply]
Okay thanks for your support

**Reply by tarsil** [reply]
Not a problem. Now everything should be ok

**Comment by saurabhgharatpara** [comment]
ive added q: Dict[str, str] = None as an arument for query params . query params is optional
but im getting this error 
![capt-2-20-8-2024](https://github.com/user-attachments/assets/c306d16e-6e87-41e9-9ea2-8e7f6af63009)
![capt-20-8-2024](https://github.com/user-attachments/assets/1418f73b-5842-4e63-86e0-4cc7ffeeb596)

  this is my function 
  async def test(request: Request, headers: dict, q: Dict[str, str] = None, 
                                  **kwargs: Any) -> Response:

**Comment by saurabhgharatpara** [comment]
This issue is solved by making changes in signature.py file for parse_values_for_connection function 

def parse_values_for_connection(cls, connection: Union[Request, WebSocket], **kwargs: Dict[str, Any]) -> Any:
        """
        Parses keyword arguments for connection, applies encoders if defined,
        and retrieves validated model fields.

        Args:
            connection (Union[Request, WebSocket]): The connection object.
            kwargs (Any): Keyword arguments to parse.

        Returns:
            Dict[str, Any]: Dictionary of validated model fields.

        Raises:
            BaseSystemException: If validation error occurs.
            EncoderException: If encoder error occurs.
        """
        try:
            if cls.encoders:
                kwargs = cls.parse_encoders(kwargs)
            if not kwargs.get("q"):
                kwargs["q"] = {}
            signature = cls(**kwargs)
            values = {}
            for key in cls.model_fields:
                values[key] = signature.field_value(key)
            return values
        except ValidationError as e:
            raise cls.build_base_system_exception(connection, e) from e
        except Exception as e:
            raise cls.build_encoder_exception(connection, e) from e

**Reply by tarsil** [reply]
@saurabhgharatpara the problem there is that you are seeing the `q` hardcoded there and this cannot happen because I might want to call `query` for example. Let me investigate

**Reply by tarsil** [reply]
@saurabhgharatpara Ok, I think I know what might be the problem, as per the documentation, if you want to make Optional, **you must  type it as such**, see [here](https://www.esmerald.dev/extras/query-params/#declaring-defaults). For example.

```python
async def test(request: Request, headers: dict, q: typing.Optional[Dict[str, str]] = None, **kwargs: Any) -> Response:
```

Or

```python
async def test(request: Request, headers: dict, q: typing.Union[Dict[str, str], None] = None, **kwargs: Any) -> Response:
```

**Reply by saurabhgharatpara** [reply]
if i make it Optional,  in swagger q is  taking value as string but i need to use dictionary which allows for dynamic key value

**Reply by tarsil** [reply]
@saurabhgharatpara Correct, there is a bug here for the swagger. Having a look and will release it soon as we are adding some extras.

**Reply by tarsil** [reply]
@saurabhgharatpara we are adding some new features and this might take some time but basically this is the OpenAPI UI that is not parsing the way it is supposed to for optional datastructures. It understands if it's mandatory as per docs and primitive typed but not mandatory and complex datastructures, it doesn't not generate the proper UI. 

We are adding some new things and we will be addressing also this small changes allowing Esmerald to be even more modular and dynamic.

**Reply by saurabhgharatpara** [reply]
Thankyou ... waiting for release soon

**Reply by saurabhgharatpara** [reply]
Hello .. any updates on this ?

**Reply by tarsil** [reply]
> Hello .. any updates on this ?

Hey. So the internal implementation for this already happened but needs some other adjustments and documentation. We expect in the next couple of weeks to do the release

**Reply by saurabhgharatpara** [reply]
Okay..thank you for the update

**Reply by tarsil** [reply]
@saurabhgharatpara we found a possible bug in the OpenAPI representation and we will be doing a release now for other purposes (multi body) while we investigate further this but don't worries as we intend to make this clearer in a new release as well later with the query params.

**Comment by tarsil** [comment]
@saurabhgharatpara apologies for the time it took but we needed to investigate the source cause for this and it wasn't a simple thing. Not a bug from Esmerald but from one of the dependencies. This is now working 100% ok with the new version released just now. 3.4.1

Have a look at the release notes but it should now work like a charm for optional query params.

**Comment by tarsil** [comment]
@saurabhgharatpara i received an email with a reply from you but I can't find it in the conversation. Did you delete it?

**Reply by saurabhgharatpara** [reply]
yes

**Reply by tarsil** [reply]
So, it works?

**Reply by saurabhgharatpara** [reply]
Yes

**Reply by tarsil** [reply]
Good, case closed then.

**Reply by saurabhgharatpara** [reply]
Hello, im facing this issue in 3.4.1
Im getting this issue in swagger input should be valid dictionary
![capt-7-10-24](https://github.com/user-attachments/assets/209a641c-251c-46a8-bdae-827b6909b870)
![cpt-7-10-24-2](https://github.com/user-attachments/assets/b2973505-9066-4482-9288-ebd816562c06)
![cpt-7-10-24-3](https://github.com/user-attachments/assets/57c2221a-255c-457d-be38-d404a5496db1)

**Reply by tarsil** [reply]
> Hello, im facing this issue in 3.4.1 Im getting this issue in swagger input should be valid dictionary ![capt-7-10-24](https://private-user-images.githubusercontent.com/176920935/374052734-209a641c-251c-46a8-bdae-827b6909b870.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjgyODcyNTEsIm5iZiI6MTcyODI4Njk1MSwicGF0aCI6Ii8xNzY5MjA5MzUvMzc0MDUyNzM0LTIwOWE2NDFjLTI1MWMtNDZhOC1iZGFlLTgyN2I2OTA5Yjg3MC5QTkc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDA3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAwN1QwNzQyMzFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05MDg4MWNmZjkwMWM1YjY5NjRjMzYyZTlkMDhiNzE1YmY3NjU2MWQ5Mzk3YTQ4NmFhZDVhYmJlNDQzZDJmMmYzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.i278CTMAbtU-5YTmZlPTUmO0Nybpqbn4MerB7TTxyBk) ![cpt-7-10-24-2](https://private-user-images.githubusercontent.com/176920935/374052741-b2973505-9066-4482-9288-ebd816562c06.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjgyODcyNTEsIm5iZiI6MTcyODI4Njk1MSwicGF0aCI6Ii8xNzY5MjA5MzUvMzc0MDUyNzQxLWIyOTczNTA1LTkwNjYtNDQ4Mi05Mjg4LWViZDgxNjU2MmMwNi5QTkc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDA3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAwN1QwNzQyMzFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iNTMzYWFiYWQxMDNhZDIyOTQwZjQwODRmMTJmYjVjZmQ5OTcwYjRjNjI3ZDc5MTc5ZGI3MjE4YjViOWQxYjVhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.0G0cBAUV7yiz-lTqJv88eOiSli8tMDIKVirn_dvlMtg) ![cpt-7-10-24-3](https://private-user-images.githubusercontent.com/176920935/374052745-57c2221a-255c-457d-be38-d404a5496db1.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjgyODcyNTEsIm5iZiI6MTcyODI4Njk1MSwicGF0aCI6Ii8xNzY5MjA5MzUvMzc0MDUyNzQ1LTU3YzIyMjFhLTI1NWMtNDU3ZC1iZTM4LWQ0MDRhNTQ5NmRiMS5QTkc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDA3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAwN1QwNzQyMzFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xYmQ5ZWZlMGI2ZGEyMDg3ODMxZjQwMmU0Yjk3N2M4ZWE1NjJkMjhjNjdkNzczY2JkZTc2ZDg2MmExMTkzY2RkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.6udUYp9h31h-AqUvYmI3Tbb2wSax5ORZF5j9u9huDg4)

Thank you for this. I'm currently using Esmerald 3.4.2 (the latest) and I literally replicated this and I am not having issues at all.


![image](https://github.com/user-attachments/assets/d495cc38-9e81-41d9-a55b-d5c43ca10639)
![image](https://github.com/user-attachments/assets/261367e1-41d0-434d-a247-ee3f02da694b)


No issues at all

**Reply by saurabhgharatpara** [reply]
Yes ive upgraded to latest version . its working Thanks

**Reply by tarsil** [reply]
> Yes ive upgraded to latest version . its working Thanks

I found it weird as I didn't have that problem also with the version 3.4.1. Anyway, I'm glad it's addressed
